src/components/KTableData/KTableData.vue

```diff
@@ -172,91 +170,90 @@ import type {
-  errorStateMessage: 'Data cannot be displayed due to an error.',
-  errorStateActionMessage: '',
-  maxHeight: 'none',
-  hidePagination: false,
-  paginationAttributes: () => ({}),
-  rowExpandable: () => false,
-  rowExpanded: () => false,
-  hideHeaders: false,
-  nested: false,
-  hidePaginationWhenOptional: false,
-  hideToolbar: false,
-  /**
-   * KTableData props defaults
-   */
-  headers: () => ([]),
-  fetcherCacheKey: '',
-  cacheIdentifier: '',
-  searchInput: '',
-  initialFetcherParams: () => ({}),
-  clientSort: false,
-  sortable: true,
-})
+type ColumnKey = TableColumnKey<Header>
+type ColumnVisibility = TableColumnVisibility<Header>
+type ColumnWidths = TableColumnWidths<Header>
 
-const slots = useSlots()
+const {
+  resizeColumns,
+  tablePreferences = {},
+  rowHover = true,
+  rowAttrs = () => ({}),
+  rowLink = () => ({} as RowLink),
+  rowBulkActionEnabled = () => true,
+  rowKey: rowKeyProp,
+  cellAttrs = () => ({}),
+  loading,
+  emptyStateTitle = 'No Data',
+  emptyStateMessage = 'There is no data to display.',
+  emptyStateActionMessage = '',
+  emptyStateIconVariant = EmptyStateIconVariants.Default,
+  emptyStateButtonAppearance,
+  emptyStateActionRoute,
+  error,
+  errorStateTitle = 'An error occurred',
+  errorStateMessage = 'Data cannot be displayed due to an error.',
+  errorStateActionMessage = '',
+  errorStateActionRoute,
+  maxHeight = 'none',
+  hidePagination,
+  paginationAttributes = {},
+  rowExpandable = () => false,
+  rowExpanded = () => false,
+  hideHeaders,
+  nested,
+  hidePaginationWhenOptional,
+  hideToolbar,
+  headers = [],
+  fetcherCacheKey = '',
+  cacheIdentifier = '',
+  searchInput = '',
+  initialFetcherParams = {},
+  clientSort,
+  sortable = true,
+  fetcher,
+  sortHandlerFunction,
+  tooltipTarget,
+  ...restProps
+} = defineProps<TableDataProps<Header, Data, Offset>>()
+
+const emit = defineEmits<TableDataEmits<Header, Data>>()
+
+const slots = defineSlots<TableDataSlots<Header, Data>>()
 
 const { useDebounce, useRequest, useSwrvState, clientSideSorter: defaultClientSideSorter } = useUtilities()
 
-const emit = defineEmits<{
-  (e: 'cell-click', value: { data: Record<string, any> }): void
-  (e: 'row-click', value: { data: Record<string, any> }): void
-  (e: 'error-action-click'): void
-  (e: 'empty-state-action-click'): void
-  (e: 'update:table-preferences', preferences: TablePreferences): void
-  (e: 'sort', value: TableSortPayload): void
-  (e: 'state', value: TableStatePayload): void
-  (e: 'row-select', data: Record<string, any>[]): void
-  (e: 'update:row-expanded', data: RowExpandPayload): void
-  (e: 'row-actions-toggle', data: RowActionsTogglePayload): void
-}>()
-
 const tableId = useId()
 
-const tableData = ref<Record<string, any>[]>([])
-const tableHeaders = computed((): TableDataHeader[] => props.sortable ? props.headers : props.headers.map((header) => ({ ...header, sortable: false })))
+const tableData = ref([]) as Ref<Data[number][]>
```

Comment:
nit: it might be useful to leave comments in the PR as to why we are casting so that it's clear for future devs?

Reply:
Added.